pkg/storage/tsdb/ref_cache.go

```diff
@@ -0,0 +1,161 @@
+package tsdb
+
+import (
+	"sync"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+)
+
+const (
+	// DefaultRefCacheTTL is the default RefCache purge TTL. We use a reasonable
+	// value that should cover most use cases. The cache would be ineffective if
+	// the scrape interval of a series is greater than this TTL.
+	DefaultRefCacheTTL = int64(5 * time.Minute)
+
+	refCacheStripes = 128
+)
+
+// RefCache is a single-tenant cache mapping a labels set with the reference
+// ID in TSDB, in order to be able to append samples to the TSDB head without having
+// to copy write request series labels each time (because the memory buffers used to
+// unmarshal the write request is reused).
+type RefCache struct {
+	// The cache is splitted into stripes, each one with a dedicated lock, in
+	// order to reduce lock contention.
+	stripes map[uint8]*refCacheStripe
+}
+
+// refCacheStripe holds a subset of the series references for a single tenant.
+type refCacheStripe struct {
+	refsMu sync.Mutex
+	refs   map[model.Fingerprint][]*refCacheEntry
+}
+
+// refCacheEntry holds a single series reference.
+type refCacheEntry struct {
+	lbs       labels.Labels
+	ref       uint64
+	touchedAt int64
+}
+
+// NewRefCache makes a new RefCache.
+func NewRefCache() *RefCache {
+	c := &RefCache{
+		stripes: make(map[uint8]*refCacheStripe, refCacheStripes),
+	}
+
+	// Stripes are pre-allocated so that we only read on them and no lock is required.
+	for i := uint8(0); i < refCacheStripes; i++ {
+		c.stripes[i] = &refCacheStripe{
+			refs: map[model.Fingerprint][]*refCacheEntry{},
+		}
+	}
+
+	return c
+}
+
+// GetRef returns the cached series reference, and guarantees the input labels set
+// is NOT retained.
+func (c *RefCache) GetRef(now int64, series []labels.Label) (uint64, bool) {
+	fp := client.FastFingerprint(client.FromLabelsToLabelAdapters(series))
+
+	// Get the related stripe
+	stripeID := uint8(util.HashFP(fp) % refCacheStripes)
```

Comment:
related to above about changing from map -> array, since we already hash and modulo we should use an array instead.